src/wallet/rpcdump.cpp

```diff
@@ -691,31 +691,20 @@ UniValue importwallet(const JSONRPCRequest& request)
             const CScript& script = script_pair.first;
             int64_t time = script_pair.second;
             CScriptID id(script);
-            if (pwallet->HaveCScript(id)) {
-                pwallet->WalletLogPrintf("Skipping import of %s (script already present)\n", HexStr(script));
-                continue;
-            }
-            if(!pwallet->AddCScript(script)) {
-                pwallet->WalletLogPrintf("Error importing script %s\n", HexStr(script));
-                fGood = false;
+
+            if (pwallet->ImportScripts({script}, time)) {
+                pwallet->WalletLogPrintf("Skipping import of %s (script already present or an error occurred)\n", HexStr(script));
                 continue;
             }
-            if (time > 0) {
-                pwallet->m_script_metadata[id].nCreateTime = time;
-                nTimeBegin = std::min(nTimeBegin, time);
-            }
+
             progress++;
         }
         pwallet->chain().showProgress("", 100, false); // hide progress dialog in GUI
-        pwallet->UpdateTimeFirstKey(nTimeBegin);
     }
     pwallet->chain().showProgress("", 100, false); // hide progress dialog in GUI
     RescanWallet(*pwallet, reserver, nTimeBegin, false /* update */);
     pwallet->MarkDirty();
 
-    if (!fGood)
-        throw JSONRPCError(RPC_WALLET_ERROR, "Error adding some keys/scripts to wallet");
```

Comment:
e66ed3b84593bebd3550e18b5b3be3605e08ced1

These changes are kind of breaking change, the caller has no way to know if something went bad - as it is will always succeed.Is this the case just because `dumpwallet`/`importwallet` should not be used?

Reply:
This change was because it `ImportScripts` and `ImportPrivKeys` doesn't return why it failed. It could fail because the script or privkey already exists in the wallet which would still be fine, or it could cail because something went wrong with writing the data to the wallet file which is bad. I suppose I could make them return an enum, but that's a larger change that I didn't really want to do.

Reply:
The errors reported as-is appear to at least be somewhat helpful.

> It could fail because the script or privkey already exists in the wallet 

I don't think that's true. These two checks appear to be only reporting real errors. Perhaps I missed something.

Reply:
Yes, but these checks are part of `ImportPrivKeys` and `ImportScripts` both of which can return false in non-error cases. Both have `HaveKey` and/or `HaveCScript` checks that can fail, but not cause the import to be bad. But there could also be an actual error which would result in the import being bad.

Reply:
In commit "Have importwallet use ImportPrivKeys and ImportScripts" (991ecfdbbb289bd6e4c50fc63f9201edbca14f70)

> Both have HaveKey and/or HaveCScript checks that can fail, but not cause the import to be bad. 

What's this referring to? Import methods seem to return true if imported data was already found. But maybe I'm misinterpreting. Might help to refer to a specific example.

Reply:
Turns out I was having a hard time with the negations.

I've pushed another commit that makes it more explicit when something is being skipped because we already have it, as well as logging when this happens.